src/avro_util.erl

```diff
@@ -295,6 +296,107 @@ resolve_duplicated_refs(Type0) ->
   {Type, _Refs} = resolve_duplicated_refs(Type0, []),
   Type.
 
+%% @doc Check if writer schema is compatible to reader schema.
+-spec is_compatible(avro_type(), avro_type()) ->
+                       true | {not_compatible, _, _} |
+                       {reader_missing_defalut_value, [{field, _} | _]}.
+is_compatible(Reader, Writer) ->
+  try
+    do_is_compatible_next(Reader, Writer, [], [])
+  catch
+    throw : {not_compatible, RPath, WPath} ->
+      {not_compatible, lists:reverse(RPath), lists:reverse(WPath)};
+    throw : {reader_missing_defalut_value, Path} ->
+      {reader_missing_default_value, lists:reverse(Path)}
+  end.
+
+do_is_compatible_next(Reader, Writer, RPath, WPath) ->
+  NewRPath = [avro:get_type_fullname(Reader) | RPath],
+  NewWPath = [avro:get_type_fullname(Writer) | WPath],
+  do_is_compatible(Reader, Writer, NewRPath, NewWPath).
+
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_RECORD_TYPE(Reader) andalso ?IS_RECORD_TYPE(Writer) ->
+  SameType = avro:is_same_type(Reader, Writer),
+  ReaderTypes = avro_record:get_all_field_data(Reader),
+  WriterTypes = avro_record:get_all_field_data(Writer),
+  SameType andalso
+    lists:all(
+      fun({FieldName, FieldType, Default}) ->
+          case lists:keysearch(FieldName, 1, WriterTypes) of
+            false ->
+              Default =/= ?NO_VALUE orelse
+                erlang:throw({ reader_missing_defalut_value
+                             , [{field, FieldName} |  WPath]
+                             });
+            {value, {_, WriterType, _}} ->
+              FieldDesc = {field, FieldName},
+              do_is_compatible_next(FieldType, WriterType,
+                               [FieldDesc | RPath],
+                               [FieldDesc | WPath])
+          end
+      end,
+      ReaderTypes);
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_ENUM_TYPE(Reader) andalso ?IS_ENUM_TYPE(Writer) ->
+  avro:is_same_type(Reader, Writer) andalso
+    Writer#avro_enum_type.symbols -- Reader#avro_enum_type.symbols == [] orelse
+    erlang:throw({not_compatible, RPath, WPath});
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_FIXED_TYPE(Reader) andalso ?IS_FIXED_TYPE(Writer) ->
+  avro:is_same_type(Reader, Writer)
+    andalso (Reader#avro_fixed_type.size == Writer#avro_fixed_type.size) orelse
+    erlang:throw({not_compatible, RPath, WPath});
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_UNION_TYPE(Reader) andalso ?IS_UNION_TYPE(Writer) ->
+  WriterTypes = avro_union:get_types(Writer),
+  lists:all(
+    fun(WriterType) ->
+        do_is_compatible_next(Reader, WriterType, tl(RPath), WPath)
+    end, WriterTypes);
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_UNION_TYPE(Reader) ->
+  ReaderTypes = avro_union:get_types(Reader),
+  ReaderTypeCount = length(ReaderTypes),
+  ReaderTypeIndices = lists:zip(lists:seq(0, ReaderTypeCount - 1), ReaderTypes),
```

Comment:
a second thought about member_id:
since same type can not appear more than once in a union,
and we always add type name to paths,
there is actually no need to id them.